### ISSUE TITLE
Creating Observable with back pressure from asynchronous message queue

### ISSUE THREAD

**Issue by fancywriter** [monix/monix#691]
Hello! I am reading monix Observable documentation and not sure that I understand everything I need.
My use case: I have message queue (suppose rabbit MQ, but shouldn't really matter, suppose I use amqp-client). I want to just read everything from the queue and write to the local file. And I want this process to be easily cancelled and stopped with closing all resources correctly (flushing buffer to the file).

I have two options with AMQP client, I have `basicConsume` method, which is asynchronous and `basicGet`, which is synchronous. Of course I want to use first option. Although it might be easier to implement back pressure with `basicGet`.

How should I create observable for this? Obviously I can't use `Observable.create` doesn't support back pressure (if I understand documentation correctly)? But using `new Observable` and overriding `unsafeSubscribeFn` seems to be unsafe. What is the proper way?

And do I understand correctly that AMQP itself doesn't allow me to do back pressure at all? It doesn't allow me to pause consuming, at least I don't see easy way to do it, only to cancel it.

**Comment by Avasil** [comment]
Maybe you can use `Task` to get result from `basicConsume` if it's asynchronous and create `Observable` with `Observable.fromTask`?

Don't know about backpressure in AMQP part because it has been a while since I was doing anything with RabbitMQ but I remember having options like maximum unacked messages

**Comment by Executioner1939** [comment]
@fancywriter I have implemented something similar, I relied on AMQP to do the back-pressuring using the [Consumer Prefetch (Qos)](https://www.rabbitmq.com/consumer-prefetch.html).

I then use ConcurrentSubject  like this:

```
val source = ConcurrentSubject.publish[Delivery[Array[Byte]]](OverflowStrategy.DropNew(100)).doOnNextAck {
    case (delivery, Continue) => channel.basicAck
    case (delivery, Stop) => // Error Code
}

// Called in handleDelivery
source.onNext(delivery)
```
So that is one way to do it :-). That way you get an Observable with Rabbit doing Back-Pressure so you don't need to worry about it :-).

Hopefully that is helpful, it's just a suggestion, not necessarily correct at all :-).
